### PR TITLE
[NTUSER] Add ASSERT() on "linking window to itself"

### PR DESCRIPTION
**Pull request by SergeGautherie**
## Purpose

@learn-more wrote 'break on this in debug builds'.
https://jira.reactos.org/browse/CORE-18132?focusedCommentId=133294&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-133294

Addendum to ee0511b (0.4.10-dev-323).
JIRA issue: [CORE-18132](https://jira.reactos.org/browse/CORE-18132)

## Proposed changes

- Add 1 `else`.
- Add 2 `ASSERT()`.
- Add 1 FIXME comment.

### CODE REVIEW

File: win32ss/user/ntuser/window.c [hunk 1]
```diff
@@ -937,7 +937,8 @@ IntLinkWindow(
 {
     if (Wnd == WndInsertAfter)
     {
-        ERR("IntLinkWindow -- Trying to link window 0x%p to itself!!\n", Wnd);
+        ERR("%s -- Trying to link window 0x%p to itself!!\n", __FUNCTION__, Wnd);
```
Comment (HBelusca) [inline win32ss/user/ntuser/window.c:940]: Don't do more changes than necessary.
```suggestion
        ERR("IntLinkWindow -- Trying to link window 0x%p to itself!!\n", Wnd);
```

File: win32ss/user/ntuser/window.c [hunk 2]
```diff
@@ -1040,8 +1041,15 @@ VOID FASTCALL IntLinkHwnd(PWND Wnd, HWND hWndPrev)
         }
 
         if (Wnd == WndInsertAfter)
-            ERR("IntLinkHwnd -- Trying to link window 0x%p to itself!!\n", Wnd);
-        IntLinkWindow(Wnd, WndInsertAfter);
+        {
+            ERR("%s -- Trying to link window 0x%p to itself!!\n", __FUNCTION__, Wnd);
```
Comment (HBelusca) [inline win32ss/user/ntuser/window.c:1045]:
```suggestion
            ERR("IntLinkHwnd -- Trying to link window 0x%p to itself!!\n", Wnd);
```

File: win32ss/user/ntuser/window.c [hunk 3]
```diff
@@ -1040,8 +1041,15 @@ VOID FASTCALL IntLinkHwnd(PWND Wnd, HWND hWndPrev)
...
+            ASSERT(WndInsertAfter != Wnd);
+            FIXME("IntUnlinkWindow() was already called. Continuing as is seems wrong!");
```
Comment (HBelusca) [inline win32ss/user/ntuser/window.c:1047]: Looks to me like it's in a sense a similar case to the one above "Are we called with an erroneous handle" :
there reasoning here being that the given `WndInsertAfter` is "wrong", too. Not as a value itself, but by its relation with `Wnd` .

File: win32ss/user/ntuser/window.c (hunk 1 above)
Comment (SergeGautherie) [reply win32ss/user/ntuser/window.c:940]: My main idea is that identical strings can be merged, saving some bits of memory.

Comment (JoachimHenze) [reply win32ss/user/ntuser/window.c:940]: I don't have a strong opinion regarding your fight, but IF you intend to change those lines, then remove one or two exclamation marks at the end of the sentence also.

File: win32ss/user/ntuser/window.c (hunk 3 above)
Comment (JoachimHenze) [reply win32ss/user/ntuser/window.c:1047]: same here. The exclamation mark is unnecessary. The log is full of them anyway. The several loggings fighting for the longest dick in this regard is a terrible development in every codebase.

File: win32ss/user/ntuser/window.c (hunk 2 above)
Comment (JoachimHenze) [reply win32ss/user/ntuser/window.c:1045]: see my proposal above.

File: win32ss/user/ntuser/window.c [hunk 4]
```diff
@@ -1040,8 +1041,15 @@ VOID FASTCALL IntLinkHwnd(PWND Wnd, HWND hWndPrev)
         }
 
         if (Wnd == WndInsertAfter)
-            ERR("IntLinkHwnd -- Trying to link window 0x%p to itself!!\n", Wnd);
-        IntLinkWindow(Wnd, WndInsertAfter);
+        {
+            ERR("Trying to link window 0x%p to itself\n", Wnd);
+            ASSERT(WndInsertAfter != Wnd);
+            FIXME("IntUnlinkWindow() was already called. Continuing as is seems wrong\n");
```
Comment (JoachimHenze) [inline win32ss/user/ntuser/window.c:1047]: Placing the FIXME after the assert doesn't seem to make a lot of sense to me. Why not adding that text to the ERR() above?

Comment (learn-more) [reply win32ss/user/ntuser/window.c:1047]: I am not sure about this fixme, but the rest lgtm

**Review by learn-more — APPROVED** [review]

File: win32ss/user/ntuser/window.c (hunk 4 above)
Comment (SergeGautherie) [reply win32ss/user/ntuser/window.c]: I demoted the FIXME line to a comment. That should be enough for this PR.

File: win32ss/user/ntuser/window.c (hunk 3 above)
Comment (SergeGautherie) [reply win32ss/user/ntuser/window.c]: @HBelusca, can this case even happen, after `IntUnlinkWindow(Wnd)`?

> "Are we called with an erroneous handle"

If yes, then would a better fix be to check that `UserGetWindowObject(hWndPrev) != Wnd` at the start of current function, instead?

**Review by HBelusca — APPROVED** [review]

**Review by JoachimHenze — APPROVED** [review]

**Pull request merged**